File: App/src/Adaptive.ReactiveTrader.Client.Domain/Models/Price.cs
using System;

namespace Adaptive.ReactiveTrader.Client.Domain.Models
{
    class Price : IPrice
    {
        public Price(ExecutablePrice bid, ExecutablePrice ask, long quoteId, DateTime valueDate, ICurrencyPair currencyPair)
        {
            Bid = bid;
            Ask = ask;
            QuoteId = quoteId;
            ValueDate = valueDate;
            CurrencyPair = currencyPair;

            bid.Parent = this;
            ask.Parent = this;

            Spread = (ask.Rate - bid.Rate)* (long)Math.Pow(10, currencyPair.PipsPosition);
        }

        public IExecutablePrice Bid { get; private set; }
        public IExecutablePrice Ask { get; private set; }
        public ICurrencyPair CurrencyPair { get; private set; }
        public long QuoteId { get; private set; }
        public DateTime ValueDate { get; private set; }
        public decimal Spread { get; private set; }
        public bool IsStale { get { return false; } }
    }
}

File: src/Adaptive.ReactiveTrader.Client.JS/Models/Pricing/Price.ts
class Price implements IPrice, IPriceLatency {
    constructor(
        bid: ExecutablePrice,
        ask: ExecutablePrice,
        mid: number,
        quoteId: number,
        valueDate: Date,
        currencyPair: ICurrencyPair) {

        this.bid = bid;
        this.ask = ask;
        this.mid = mid;
        this.quoteId = quoteId;
        this.valueDate = valueDate;
        this.currencyPair = currencyPair;
        this.isStale = false;

        bid.parent = this;
        ask.parent = this;

        this.spread = (ask.rate - bid.rate) * Math.pow(10, currencyPair.pipsPosition);
        this._receivedTimestamp = performance.now();
    }

    // TODO encapsulate as fields
    bid: IExecutablePrice;
    ask: IExecutablePrice;
    mid: number; 
    currencyPair: ICurrencyPair;
    quoteId: number;
    valueDate: Date;
    spread: number;
    isStale: boolean; 

    // IPriceLatency implementation

    private _receivedTimestamp: number;
    private _renderTimestamp: number;

    get uiProcessingTimeMs() {
        return this._renderTimestamp - this._receivedTimestamp;
    }

    displayedOnUi(): void {
        this._renderTimestamp = performance.now();
    }
} 